README.md
### Purpose
The default M2 sample data does not contain a multi-site solution, this introduces some data to work with
when developing tests. Ideally we would also create other sample data modules which would introduce categories,
products, etc. to the websites here. It would be bare minimum to increase speed of installation and running. 

### Structure
* web_computer - default, USD
    * grp_desktop
        * store_desktop_en - English, lbs
        * store_desktop_ru - Russian, kgs
        * store_desktop_de - German, kgs, default
        * store_desktop_dis - English, lbs, disabled
    * grp_laptop - default
        * store_laptop_en - English, lbs, default
        * store_laptop_ru - Russian, kgs
        * store_laptop_de - German, kgs
        * store_laptop_dis - English, lbs, disabled
* web_phone - RUB
    * grp_tablet - default
        * store_tablet_en - English, lbs
        * store_tablet_ru - Russian, kgs default
        * store_tablet_de - German, kgs
        * store_tablet_dis - English, lbs, disabled
    * grp_accessory - disabled
        * store_accessory_en - English, lbs, default, disabled
        * store_accessory_ru - Russian, kgs, disabled
        * store_accessory_de - German, kgs, disabled

### Installation
Can be installed manually by including the package via composer:

```$xslt
composer require shopgate/magento2-website-sample-data
```

If you have any other magento 2 plugin that suggests this sample data, e.g. [shopgate/connect-integration-magento2] 
the sample data will be installed via regular magento command `bin/magento sampledata:deploy`

### Tests
Run tests the same way you do normally. You may also adjust your integration phpunit.xml to include the tests in 
the run.
```$xslt
 /var/www/html/vendor/phpunit/phpunit/phpunit --configuration /var/www/html/dev/tests/integration/phpunit.xml /var/www/html/vendor/shopgate/magento2-website-sample-data/Test/Integration
```

### Uninstall
The best way to remove the module fully is manually:
```$xslt
bin/magento module:uninstall Shopgate_WebsiteSampleData --remove-data
```

You could also use the standard removal, but unlike the previous solution there is no hook that will remove the
 websites, groups, stores and configs from the database. Just the module.
```$xslt
bin/magento sampledata:remove
bin/magento setup:upgrade
```

### Todo's
We'll need to populate these entities with proper data to do tests on.

* Create categories
* Create products

[shopgate/connect-integration-magento2]: https://github.com/shopgate/connect-integration-magento2


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [0.2.1]
### Updated
- M2 composer dependencies (to support PHP8.2 & M2.4.6)

## [0.2.0]
### Added
- Currency configurations
- Default website, groups and store views
- Thorough tests
- Removal feature

## [0.1.0]
### Added
- 2 websites
- 5 store groups
- 15 store views
- Language configurations
- Weight configurations
- Basic test

[Unreleased]: https://github.com/shopgate/magento2-website-sample-data/compare/0.2.0...HEAD
[0.2.0]: https://github.com/shopgate/magento2-website-sample-data/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/shopgate/magento2-website-sample-data/releases/tag/0.1.0


Test/Integration/WebsiteSampleData/Setup/InstallerTest.php
<?php

/**
 * Copyright Shopgate Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @author    Shopgate Inc, 804 Congress Ave, Austin, Texas 78701 <interfaces@shopgate.com>
 * @copyright Shopgate Inc
 * @license   http://www.apache.org/licenses/LICENSE-2.0 Apache License, Version 2.0
 */

declare(strict_types=1);

namespace Shopgate\WebsiteSampleData\Test\Integration\Setup;

use Magento\Directory\Helper\Data as DirectoryHelper;
use Magento\Directory\Model\Currency;
use Magento\Framework\App\Config\ScopeConfigInterface;
use Magento\Framework\Exception\LocalizedException;
use Magento\Store\Model\ScopeInterface;
use Magento\Store\Model\StoreManagerInterface;
use Magento\TestFramework\Helper\Bootstrap;
use Magento\TestFramework\ObjectManager;
use PHPUnit\Framework\TestCase;
use Shopgate\WebsiteSampleData\Setup\Installer;

/**
 * @magentoDbIsolation  enabled
 * @magentoAppArea      adminhtml
 */
class InstallerTest extends TestCase
{
    /** @var ObjectManager $objectManager */
    protected $objectManager;

    /**
     * Basic setup
     */
    public function setUp()
    {
        $this->objectManager = Bootstrap::getObjectManager();
    }

    /**
     * @throws LocalizedException
     */
    public function testInstall(): void
    {
        /** @var Installer $installer */
        $installer = $this->objectManager->get(Installer::class);
        $installer->install();

        /** @var StoreManagerInterface $storeManager */
        $storeManager = $this->objectManager->get(StoreManagerInterface::class);
        $this->assertCount(16, $storeManager->getStores());
        $this->assertCount(3, $storeManager->getWebsites());
        $this->assertCount(5, $storeManager->getGroups());
        $laptopStore = $storeManager->getStore('store_laptop_de');
        /** @var ScopeConfigInterface $reader */
        $reader            = $this->objectManager->get(ScopeConfigInterface::class);
        $laptopStoreLocale = $reader->getValue(
            DirectoryHelper::XML_PATH_DEFAULT_LOCALE,
            ScopeInterface::SCOPE_STORES,
            $laptopStore->getId()
        );
        $this->assertSame('de_DE', $laptopStoreLocale);

        $phoneWebsite         = $storeManager->getWebsite('web_phone');
        $phoneWebsiteCurrency = $reader->getValue(
            Currency::XML_PATH_CURRENCY_BASE,
            ScopeInterface::SCOPE_WEBSITES,
            $phoneWebsite->getId()
        );
        $this->assertSame('RUB', $phoneWebsiteCurrency);
    }
}


Test/Integration/WebsiteSampleData/Setup/UninstallerTest.php
<?php

/**
 * Copyright Shopgate Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @author    Shopgate Inc, 804 Congress Ave, Austin, Texas 78701 <interfaces@shopgate.com>
 * @copyright Shopgate Inc
 * @license   http://www.apache.org/licenses/LICENSE-2.0 Apache License, Version 2.0
 */

declare(strict_types=1);

namespace Shopgate\WebsiteSampleData\Test\Integration\Setup;

use Magento\Directory\Helper\Data as DirectoryHelper;
use Magento\Framework\App\Config\ScopeConfigInterface;
use Magento\Framework\Exception\LocalizedException;
use Magento\Framework\Setup\SchemaSetupInterface;
use Magento\Setup\Model\ModuleContext;
use Magento\Store\Model\ScopeInterface;
use Magento\Store\Model\StoreManagerInterface;
use Magento\TestFramework\Helper\Bootstrap;
use Magento\TestFramework\ObjectManager;
use PHPUnit\Framework\TestCase;
use Shopgate\WebsiteSampleData\Setup\Installer;
use Shopgate\WebsiteSampleData\Setup\Uninstall;

/**
 * @magentoDbIsolation  enabled
 * @magentoAppArea      adminhtml
 */
class UninstallerTest extends TestCase
{
    /** @var ObjectManager $objectManager */
    protected $objectManager;

    /**
     * Basic setup
     */
    public function setUp()
    {
        $this->objectManager = Bootstrap::getObjectManager();
    }

    /**
     * Don't simplify the expected exception
     *
     * @throws LocalizedException
     * @noinspection PhpFullyQualifiedNameUsageInspection
     * @expectedException \Magento\Framework\Exception\NoSuchEntityException
     */
    public function testUninstall(): void
    {
        /** @var Installer $installer */
        $installer = $this->objectManager->get(Installer::class);
        $installer->install();

        /** @var StoreManagerInterface $storeManager */
        $storeManager = $this->objectManager->get(StoreManagerInterface::class);
        $stores       = $storeManager->getStores();
        $this->assertCount(3, $storeManager->getWebsites());
        $this->assertCount(5, $storeManager->getGroups());
        $this->assertCount(16, $stores);

        /** @var Uninstall $uninstall */
        $uninstall = $this->objectManager->get(Uninstall::class);
        $setup     = $this->objectManager->get(SchemaSetupInterface::class);
        $context   = $this->objectManager->create(ModuleContext::class, ['version' => '0']);
        $uninstall->uninstall($setup, $context);

        $storeManager->reinitStores(); // clean previously loaded stores
        $this->assertCount(1, $storeManager->getWebsites());
        $this->assertCount(1, $storeManager->getGroups());
        $this->assertCount(1, $storeManager->getStores());

        // Throws no entity exception
        $reader = $this->objectManager->get(ScopeConfigInterface::class);
        $reader->getValue(
            DirectoryHelper::XML_PATH_DEFAULT_LOCALE,
            ScopeInterface::SCOPE_STORES,
            array_pop($stores)->getId()
        );

        // the default website is set again
        $websites = $storeManager->getWebsites();
        $website  = array_shift($websites);
        $this->assertSame('1', $website->getData('is_default'));
    }
}


Setup/Installer.php
<?php

/**
 * Copyright Shopgate Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @author    Shopgate Inc, 804 Congress Ave, Austin, Texas 78701 <interfaces@shopgate.com>
 * @copyright 2019 Shopgate Inc
 * @license   http://www.apache.org/licenses/LICENSE-2.0 Apache License, Version 2.0
 */

declare(strict_types=1);

namespace Shopgate\WebsiteSampleData\Setup;

use Magento\Framework\App\ObjectManager;
use Magento\Framework\Exception\LocalizedException;
use Magento\Framework\Setup;
use Magento\Store\Model\Store;
use Magento\Store\Model\StoreManagerInterface;
use Shopgate\WebsiteSampleData\Model\Website;

class Installer implements Setup\SampleData\InstallerInterface
{
    /** @var Website $website */
    protected $website;
    /** @var StoreManagerInterface */
    private $storeManager;

    /**
     * @param Website               $website
     * @param StoreManagerInterface $storeManager
     */
    public function __construct(Website $website, StoreManagerInterface $storeManager = null)
    {
        $this->website      = $website;
        $this->storeManager = $storeManager ? : ObjectManager::getInstance()->get(StoreManagerInterface::class);
    }

    /**
     * {@inheritdoc}
     * @throws LocalizedException
     */
    public function install(): void
    {
        $this->storeManager->setCurrentStore(Store::DISTRO_STORE_ID);
        $this->website->install(
            'Shopgate_WebsiteSampleData::fixtures/websites.csv',
            'Shopgate_WebsiteSampleData::fixtures/groups.csv',
            'Shopgate_WebsiteSampleData::fixtures/stores.csv',
            'Shopgate_WebsiteSampleData::fixtures/store_cfg.csv',
            'Shopgate_WebsiteSampleData::fixtures/website_cfg.csv'
        );
    }
}


Setup/Uninstall.php
<?php

/**
 * Copyright Shopgate Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @author    Shopgate Inc, 804 Congress Ave, Austin, Texas 78701 <interfaces@shopgate.com>
 * @copyright Shopgate Inc
 * @license   http://www.apache.org/licenses/LICENSE-2.0 Apache License, Version 2.0
 */

declare(strict_types=1);

namespace Shopgate\WebsiteSampleData\Setup;

use Magento\Framework\App\ResourceConnection;
use Magento\Framework\Exception\AlreadyExistsException;
use Magento\Framework\Module\Setup;
use Magento\Framework\Setup\ModuleContextInterface;
use Magento\Framework\Setup\SchemaSetupInterface;
use Magento\Framework\Setup\UninstallInterface;
use Magento\Store\Api\Data\StoreInterface;
use Magento\Store\Api\Data\WebsiteInterface;
use Magento\Store\Model\ResourceModel\Website as WebsiteResource;
use Magento\Store\Model\StoreManagerInterface;

/**
 * @codeCoverageIgnore
 */
class Uninstall implements UninstallInterface
{
    /** * @var StoreManagerInterface */
    private $storeManager;
    /** * @var WebsiteResource */
    private $websiteResource;

    /**
     * @param StoreManagerInterface $storeManager
     * @param WebsiteResource       $websiteResource
     */
    public function __construct(StoreManagerInterface $storeManager, WebsiteResource $websiteResource)
    {
        $this->storeManager    = $storeManager;
        $this->websiteResource = $websiteResource;
    }

    /**
     * Remove data that was created during module installation.
     *
     * @param SchemaSetupInterface|Setup $setup
     * @param ModuleContextInterface     $context
     *
     * @return void
     * @throws AlreadyExistsException
     */
    public function uninstall(SchemaSetupInterface $setup, ModuleContextInterface $context): void
    {
        $uninstaller       = $setup;
        $defaultConnection = $uninstaller->getConnection(ResourceConnection::DEFAULT_CONNECTION);

        $websiteCodes = ['web_computer', 'web_phone'];
        $escapedCodes = $defaultConnection->quote($websiteCodes);
        $websites     = $this->getCreatedWebsites($websiteCodes);
        $stores       = $this->getCreatedStores($websites);
        $storeIds     = $this->getKeys($stores);
        $websiteIds   = $this->getKeys($websites);
        $websiteTable = $uninstaller->getTable('store_website');
        $configTable  = $uninstaller->getTable('core_config_data');
        $defaultConnection->delete($websiteTable, "`code` IN ($escapedCodes)");
        $defaultConnection->delete($configTable, "`scope_id` IN ($storeIds) AND `scope`='stores'");
        $defaultConnection->delete($configTable, "`scope_id` IN ($websiteIds) AND `scope`='websites'");

        // set the default website back
        $websites = $this->storeManager->getWebsites();
        $website  = array_shift($websites);
        $this->websiteResource->save($website->setData('is_default', 1));
    }

    /**
     * Get stores belonging to the new websites
     *
     * @param WebsiteInterface[] $websites
     *
     * @return StoreInterface[]
     */
    public function getCreatedStores(array $websites): array
    {
        return array_map(
            static function (StoreInterface $store) use ($websites) {
                return isset($websites[$store->getWebsiteId()]);
            },
            $this->storeManager->getStores()
        );
    }

    /**
     * Get newly created websites
     *
     * @param string[] $websiteCodes
     *
     * @return WebsiteInterface[]
     */
    private function getCreatedWebsites(array $websiteCodes): array
    {
        return array_map(
            static function (WebsiteInterface $website) use ($websiteCodes) {
                return in_array($website->getCode(), $websiteCodes, true);
            },
            $this->storeManager->getWebsites()
        );
    }

    /**
     * @param WebsiteInterface[]|StoreInterface[] $entities
     *
     * @return string
     */
    private function getKeys(array $entities): string
    {
        return implode(',', array_keys($entities));
    }
}


src/registration.php
<?php

/**
 * Copyright Shopgate Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @author    Shopgate Inc, 804 Congress Ave, Austin, Texas 78701 <interfaces@shopgate.com>
 * @copyright 2019 Shopgate Inc
 * @license   http://www.apache.org/licenses/LICENSE-2.0 Apache License, Version 2.0
 */

/** @noinspection PhpFullyQualifiedNameUsageInspection */
\Magento\Framework\Component\ComponentRegistrar::register(
    \Magento\Framework\Component\ComponentRegistrar::MODULE,
    'Shopgate_WebsiteSampleData',
    __DIR__
);


Setup/Patch/Data/WebsiteSampleData.php
<?php

/**
 * Copyright Shopgate Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @author    Shopgate Inc, 804 Congress Ave, Austin, Texas 78701 <interfaces@shopgate.com>
 * @copyright 2019 Shopgate Inc
 * @license   http://www.apache.org/licenses/LICENSE-2.0 Apache License, Version 2.0
 */

declare(strict_types=1);

namespace Shopgate\WebsiteSampleData\Setup\Patch\Data;

use Magento\Framework\Setup;
use Magento\Framework\Setup\Patch\DataPatchInterface;
use Shopgate\WebsiteSampleData\Setup\Installer;

class WebsiteSampleData implements DataPatchInterface
{
    /** @var Setup\SampleData\Executor */
    protected $executor;
    /** @var Installer */
    protected $installer;

    /**
     * @param Setup\SampleData\Executor $executor
     * @param Installer                 $installer
     */
    public function __construct(Setup\SampleData\Executor $executor, Installer $installer)
    {
        $this->executor  = $executor;
        $this->installer = $installer;
    }

    /**
     * {@inheritdoc}
     */
    public function apply()
    {
        $this->executor->exec($this->installer);
    }

    /**
     * {@inheritdoc}
     */
    public static function getDependencies(): array
    {
        return [];
    }

    /**
     * {@inheritdoc}
     */
    public function getAliases(): array
    {
        return [];
    }
}


Model/Website.php
<?php

/**
 * Copyright Shopgate Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @author    Shopgate Inc, 804 Congress Ave, Austin, Texas 78701 <interfaces@shopgate.com>
 * @copyright 2019 Shopgate Inc
 * @license   http://www.apache.org/licenses/LICENSE-2.0 Apache License, Version 2.0
 */

declare(strict_types=1);

namespace Shopgate\WebsiteSampleData\Model;

use Exception;
use Magento\Directory\Helper\Data as DirectoryHelper;
use Magento\Directory\Model\Currency;
use Magento\Framework\App\Config\Storage\WriterInterface;
use Magento\Framework\Exception\AlreadyExistsException;
use Magento\Framework\Exception\LocalizedException;
use Magento\Framework\Exception\NoSuchEntityException;
use Magento\Framework\File\Csv;
use Magento\Framework\Setup\SampleData\Context as SampleDataContext;
use Magento\Framework\Setup\SampleData\FixtureManager;
use Magento\Store\Api\Data\GroupInterface;
use Magento\Store\Api\Data\StoreInterface;
use Magento\Store\Api\Data\WebsiteInterface;
use Magento\Store\Api\GroupRepositoryInterface;
use Magento\Store\Api\StoreRepositoryInterface;
use Magento\Store\Api\WebsiteRepositoryInterface;
use Magento\Store\Model\GroupFactory;
use Magento\Store\Model\ResourceModel\Group as GroupResource;
use Magento\Store\Model\ResourceModel\Store as StoreResource;
use Magento\Store\Model\ResourceModel\Website as WebsiteResource;
use Magento\Store\Model\ScopeInterface;
use Magento\Store\Model\StoreFactory;
use Magento\Store\Model\StoreManagerInterface;
use Magento\Store\Model\WebsiteFactory;

/**
 * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
 */
class Website
{
    /** @var FixtureManager */
    protected $fixtureManager;
    /** @var Csv */
    protected $csvReader;
    /** @var StoreManagerInterface */
    protected $storeManager;
    /** @var WebsiteRepositoryInterface */
    private $websiteRepository;
    /** @var GroupFactory */
    private $groupFactory;
    /** @var GroupResource */
    private $groupResource;
    /** @var GroupRepositoryInterface */
    private $groupRepository;
    /** @var WebsiteResource */
    private $webisteResource;
    /** @var WebsiteFactory */
    private $websiteFactory;
    /** @var StoreRepositoryInterface */
    private $storeRepository;
    /** @var StoreFactory */
    private $storeFactory;
    /** @var StoreResource */
    private $storeResource;
    /** * @var WriterInterface */
    private $writer;

    /**
     * @param SampleDataContext          $sampleDataContext
     * @param StoreManagerInterface      $storeManager
     * @param WebsiteFactory             $websiteFactory
     * @param WebsiteRepositoryInterface $websiteRepository
     * @param WebsiteResource            $websiteResource
     * @param GroupFactory               $groupFactory
     * @param GroupResource              $groupResource
     * @param GroupRepositoryInterface   $groupRepository
     * @param StoreFactory               $storeFactory
     * @param StoreRepositoryInterface   $storeRepository
     * @param StoreResource              $storeResource
     * @param WriterInterface            $writer
     */
    public function __construct(
        SampleDataContext $sampleDataContext,
        StoreManagerInterface $storeManager,
        WebsiteFactory $websiteFactory,
        WebsiteRepositoryInterface $websiteRepository,
        WebsiteResource $websiteResource,
        GroupFactory $groupFactory,
        GroupResource $groupResource,
        GroupRepositoryInterface $groupRepository,
        StoreFactory $storeFactory,
        StoreRepositoryInterface $storeRepository,
        StoreResource $storeResource,
        WriterInterface $writer
    ) {
        $this->fixtureManager    = $sampleDataContext->getFixtureManager();
        $this->csvReader         = $sampleDataContext->getCsvReader();
        $this->storeManager      = $storeManager;
        $this->websiteRepository = $websiteRepository;
        $this->webisteResource   = $websiteResource;
        $this->websiteFactory    = $websiteFactory;
        $this->groupFactory      = $groupFactory;
        $this->groupResource     = $groupResource;
        $this->groupRepository   = $groupRepository;
        $this->storeFactory      = $storeFactory;
        $this->storeRepository   = $storeRepository;
        $this->storeResource     = $storeResource;
        $this->writer            = $writer;
    }

    /**
     * @param string $websiteFixture
     * @param string $groupFixture
     * @param string $storeFixture
     * @param string $storeConfigFixture
     * @param string $websiteConfigFixture
     *
     * @throws LocalizedException
     */
    public function install(
        string $websiteFixture,
        string $groupFixture,
        string $storeFixture,
        string $storeConfigFixture,
        string $websiteConfigFixture
    ): void {
        $websites = $this->runCreator(
            $websiteFixture,
            function (array $row) {
                return $this->createWebsite($row['code'], $row['name'], (int) $row['is_default']);
            }
        );

        $groups = $this->runCreator(
            $groupFixture,
            function (array $row) use ($websites) {
                return $this->createGroup($row['code'], $row['name'], $websites[$row['website_code']]);
            }
        );

        $stores = $this->runCreator(
            $storeFixture,
            function (array $row) use ($websites, $groups) {
                return $this->createStore(
                    $row['code'],
                    $row['name'],
                    $websites[$row['website_code']],
                    $groups[$row['group_code']],
                    (int) $row['order'],
                    (int) $row['active']
                );
            }
        );

        // store configurations
        $this->runCreator(
            $storeConfigFixture,
            function (array $row) use ($stores) {
                $this->saveConfigs(
                    $row['locale'],
                    $row['weight_unit'],
                    (int) $stores[$row['code']]->getId()
                );
            }
        );

        // website configurations
        $this->runCreator(
            $websiteConfigFixture,
            function (array $row) use ($websites) {
                $this->writer->save(
                    Currency::XML_PATH_CURRENCY_BASE,
                    $row['currency'],
                    ScopeInterface::SCOPE_WEBSITES,
                    (int) $websites[$row['code']]->getId()
                );
            }
        );

        $this->runCreator(
            $groupFixture,
            function (array $row) use ($groups, $stores) {
                $this->updateGroupStoreDefault($groups[$row['code']], (int) $stores[$row['store_default']]->getId());
            }
        );

        $this->runCreator(
            $websiteFixture,
            function (array $row) use ($websites, $groups) {
                $this->updateWebsiteGroupDefault(
                    $websites[$row['code']],
                    (int) $groups[$row['group_default']]->getId()
                );
            }
        );
    }

    /**
     * @param string   $fixture - csv file
     * @param callable $callable
     *
     * @return string[]
     * @throws LocalizedException
     */
    private function runCreator(string $fixture, callable $callable): array
    {
        $rows   = $this->getCsv($fixture);
        $header = array_shift($rows);
        $list   = [];
        foreach ($rows as $row) {
            $data = [];
            foreach ($row as $key => $value) {
                $data[$header[$key]] = $value;
            }
            $row                = $data;
            $list[$row['code']] = $callable($row);
        }

        return $list;
    }

    /**
     * @param string $fileName
     *
     * @return string[]
     * @throws LocalizedException
     * @throws Exception
     */
    private function getCsv(string $fileName): array
    {
        $fileName = $this->fixtureManager->getFixture($fileName);

        return $this->csvReader->getData($fileName);
    }

    /**
     * @param string           $code
     * @param string           $name
     * @param WebsiteInterface $website
     * @param GroupInterface   $group
     * @param int              $order
     * @param int              $active
     *
     * @return StoreInterface
     * @throws AlreadyExistsException
     */
    private function createStore(
        string $code,
        string $name,
        WebsiteInterface $website,
        GroupInterface $group,
        int $order,
        int $active
    ): StoreInterface {

        try {
            $store = $this->storeRepository->get($code);
        } catch (NoSuchEntityException $e) {
            $store = $this->storeFactory->create();
            $store->addData(
                [
                    'code'       => $code,
                    'name'       => $name,
                    'website_id' => $website->getId(),
                    'group_id'   => $group->getId(),
                    'sort_order' => $order,
                    'is_active'  => $active
                ]
            );
            /** @noinspection PhpUnhandledExceptionInspection */
            $this->storeResource->save($store);
        }

        return $store;
    }

    /**
     * @param string $code
     * @param string $name
     * @param int    $isDefault
     *
     * @return WebsiteInterface
     * @throws AlreadyExistsException
     */
    private function createWebsite(string $code, string $name, int $isDefault): WebsiteInterface
    {
        try {
            $website = $this->websiteRepository->get($code);
        } catch (NoSuchEntityException $e) {
            $website = $this->websiteFactory->create();
            $website->addData(
                [
                    'code'       => $code,
                    'name'       => $name,
                    'is_default' => $isDefault
                ]
            );
            /** @noinspection PhpUnhandledExceptionInspection */
            $this->webisteResource->save($website);
        }

        return $website;
    }

    /**
     * @param string           $code
     * @param string           $name
     * @param WebsiteInterface $website
     *
     * @return GroupInterface
     * @throws Exception
     */
    private function createGroup(string $code, string $name, WebsiteInterface $website): GroupInterface
    {
        $groups = array_filter(
            $this->groupRepository->getList(),
            static function ($group) use ($code) {
                /** @var GroupInterface $group */
                return $group->getCode() === $code;
            }
        );
        $group  = array_pop($groups);
        if (!$group) {
            $group = $this->groupFactory->create();
            $group->addData(
                ['code' => $code, 'name' => $name, 'website_id' => $website->getId(), 'root_category_id' => 2]
            );
            $this->groupResource->save($group);
        }

        return $group;
    }

    /**
     * @param GroupInterface $group
     * @param int            $defaultStoreId
     *
     * @throws AlreadyExistsException
     */
    private function updateGroupStoreDefault(GroupInterface $group, int $defaultStoreId): void
    {
        /** @noinspection PhpParamsInspection */
        $this->groupResource->save($group->setDefaultStoreId($defaultStoreId));
    }

    /**
     * @param WebsiteInterface $website
     * @param int              $defaultGroupId
     *
     * @throws AlreadyExistsException
     */
    private function updateWebsiteGroupDefault(WebsiteInterface $website, int $defaultGroupId): void
    {
        /** @noinspection PhpParamsInspection */
        $this->webisteResource->save($website->setDefaultGroupId($defaultGroupId));
    }

    /**
     * @param string $locale - en_US, de_DE, ru_RU, etc.
     * @param string $weight - kgs, lbs
     * @param int    $id     - id of store, website or 0 for default
     * @param string $scope  - default, websites, stores
     */
    private function saveConfigs(
        string $locale,
        string $weight,
        int $id = 1,
        string $scope = ScopeInterface::SCOPE_STORES
    ): void {
        $this->writer->save(DirectoryHelper::XML_PATH_DEFAULT_LOCALE, $locale, $scope, $id);
        $this->writer->save(DirectoryHelper::XML_PATH_WEIGHT_UNIT, $weight, $scope, $id);
    }
}
